File: SourceGUI/Modules/DatabaseHelper.vb
Imports System.Data.SQLite
Imports System.IO
Imports System.Runtime.CompilerServices ' Для удобства логирования
Imports System.Data

Public Module DatabaseHelper

    ' Имя файла базы данных
    Private Const dbFileName As String = "SortHistory.sqlite"

    ' Функция для получения полного пути к файлу БД
    ' Размещает БД в той же папке, где находится .exe файл приложения
    Private Function GetDatabasePath() As String
        'Dim executablePath As String = Application.StartupPath
        'Dim projectDirectory As String = Directory.GetParent(executablePath).Parent.FullName
        Return Path.Combine(Application.StartupPath, dbFileName)
    End Function

    ' Функция для получения строки подключения
    Public Function GetConnectionString() As String
        ' Строка подключения для System.Data.SQLite
        Return $"Data Source={GetDatabasePath()}"
    End Function

    ''' <summary>
    ''' Проверяет, существует ли файл базы данных SQLite по ожидаемому пути.
    ''' </summary>
    ''' <returns>True, если файл существует, иначе False.</returns>
    ''' <remarks>
    ''' Фактическое создание файла БД обычно происходит автоматически при первом открытии соединения
    ''' (например, в функции EnsureTableExists). Эта функция полезна для проверки перед соединением.
    ''' </remarks>
    Public Function DatabaseFileExists() As Boolean
        Dim dbPath As String = GetDatabasePath()
        Try
            Return File.Exists(dbPath)
        Catch ex As Exception
            LogError($"Ошибка при проверке существования файла БД по пути '{dbPath}'.", ex)
            ' В случае ошибки проверки считаем, что файла нет (или обрабатываем иначе)
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Гарантирует, что таблица 'SortHistory' существует в базе данных.
    ''' Создает файл БД (если отсутствует) и таблицу (если отсутствует).
    ''' Эту функцию рекомендуется вызывать один раз при запуске приложения.
    ''' </summary>
    ''' <returns>True, если файл и таблица существуют или были успешно созданы, иначе False.</returns>
    Public Function EnsureTableExists() As Boolean
        Dim dbPath As String = GetDatabasePath() ' Получаем путь для логирования
        Dim fileExistedBeforeConnect As Boolean = DatabaseFileExists() ' Проверяем до соединения

        ' SQL запрос для создания таблицы, если она еще не существует
        Dim sql As String = "CREATE TABLE IF NOT EXISTS SortHistory (" &
                            "ID INTEGER PRIMARY KEY AUTOINCREMENT, " &
                            "Timestamp TEXT NOT NULL, " &
                            "OriginalArray TEXT NULL, " &
                            "SortedArray TEXT NULL);"

        Dim connectionString = GetConnectionString()
        Try
            ' Используем Using для автоматического закрытия и освобождения ресурсов
            Using conn As New SqliteConnection(connectionString)
                ' !!! Вот здесь происходит неявное создание файла БД, если его не было !!!
                conn.Open() ' Открываем соединение

                ' Логируем, если файл был только что создан
                If Not fileExistedBeforeConnect AndAlso File.Exists(dbPath) Then
                    LogInfo($"Файл базы данных '{dbPath}' не найден и был создан.")
                End If

                ' Теперь создаем таблицу в (возможно, только что созданном) файле
                Using cmd As New SqliteCommand(sql, conn)
                    cmd.ExecuteNonQuery() ' Выполняем команду создания таблицы
                End Using ' cmd освобождается
            End Using ' conn закрывается и освобождается
            Return True ' Успех
        Catch ex As Exception
            LogError("Ошибка при проверке/создании таблицы SortHistory.", ex)
            Return False ' Неудача
        End Try
    End Function

    ''' <summary>
    ''' Сохраняет исходный и отсортированный массивы в таблицу SortHistory.
    ''' </summary>
    ''' <param name="originalArray">Список целых чисел - исходный массив.</param>
    ''' <param name="sortedArray">Список целых чисел - отсортированный массив.</param>
    ''' <returns>True, если сохранение прошло успешно, иначе False.</returns>

    Public Function SaveSortResult(originalArray As List(Of Integer), sortedArray As List(Of Integer)) As Boolean
        If originalArray Is Nothing OrElse sortedArray Is Nothing Then
            LogError("Попытка сохранить пустые массивы (Nothing).")
            Return False
        End If

        ' Преобразуем списки в строки, разделенные запятыми
        Dim originalStr As String = String.Join(",", originalArray)
        Dim sortedStr As String = String.Join(",", sortedArray)
        ' Получаем текущее время и форматируем в стандарт ISO 8601 ("s")
        Dim timestampStr As String = DateTime.Now.ToString("s") ' Пример: "2025-04-07T01:27:00"

        ' SQL запрос для вставки данных с использованием параметров
        Dim sql As String = "INSERT INTO SortHistory (Timestamp, OriginalArray, SortedArray) VALUES (@Timestamp, @Original, @Sorted)"

        Dim connectionString = GetConnectionString()
        Try
            Using conn As New SQLiteConnection(connectionString)
                conn.Open()
                Using cmd As New SQLiteCommand(sql, conn)
                    ' Добавляем параметры. Это ВАЖНО для предотвращения SQL-инъекций!
                    cmd.Parameters.AddWithValue("@Timestamp", timestampStr)
                    cmd.Parameters.AddWithValue("@Original", originalStr)
                    cmd.Parameters.AddWithValue("@Sorted", sortedStr)

                    cmd.ExecuteNonQuery() ' Выполняем команду вставки
                End Using ' cmd освобождается
            End Using ' conn закрывается и освобождается
            Return True ' Успех
        Catch ex As Exception
            LogError("Ошибка при сохранении результатов сортировки в БД.", ex)
            Return False ' Неудача
        End Try
    End Function

    ''' <summary>
    '''  Получает все записи из истории сортировки, включая исходный и отсортированный массивы.
    ''' </summary>
    ''' <returns>DataTable содержащий все данные истории сортировки.</returns>
    Public Function GetAllSortHistoryData() As DataTable
        Dim historyData As New DataTable()
        Dim connectionString = GetConnectionString()

        Try
            Using conn As New SQLiteConnection(connectionString)
                conn.Open()
                ' Измененный SQL запрос для выбора всех столбцов
                Dim sql As String = "SELECT ID, Timestamp, OriginalArray, SortedArray FROM SortHistory"
                Using cmd As New SQLiteCommand(sql, conn)
                    Dim adapter As New SQLiteDataAdapter(cmd)
                    adapter.Fill(historyData)
                End Using
            End Using
        Catch ex As Exception
            LogError("Ошибка при получении данных истории сортировки.", ex)
            Return Nothing
        End Try
        Return historyData
    End Function

    ''' <summary>
    ''' Загружает исходный и отсортированный массивы для указанной записи ID.
    ''' </summary>
    ''' <param name="id">ID записи в таблице SortHistory.</param>
    ''' <returns>Кортеж (Tuple) с двумя списками List(Of Integer): (OriginalArray, SortedArray).
    ''' Возвращает Nothing, если запись не найдена или произошла ошибка.</returns>
    Public Function LoadSortResultById(id As Integer) As Tuple(Of List(Of Integer), List(Of Integer))
        Dim sql As String = "SELECT OriginalArray, SortedArray FROM SortHistory WHERE ID = @ID"
        Dim connectionString = GetConnectionString()
        Dim originalStr As String = Nothing
        Dim sortedStr As String = Nothing

        Try
            Using conn As New SQLiteConnection(connectionString)
                conn.Open()
                Using cmd As New SQLiteCommand(sql, conn)
                    cmd.Parameters.AddWithValue("@ID", id) ' Используем параметр для безопасности

                    Using reader As SQLiteDataReader = cmd.ExecuteReader()
                        If reader.Read() Then ' Если запись найдена
                            ' Получаем строки, проверяя на DBNull
                            If Not reader.IsDBNull(0) Then originalStr = reader.GetString(0)
                            If Not reader.IsDBNull(1) Then sortedStr = reader.GetString(1)
                        Else
                            LogInfo($"Запись с ID={id} не найдена в SortHistory.")
                            Return Nothing ' Запись не найдена
                        End If
                    End Using ' reader освобождается
                End Using ' cmd освобождается
            End Using ' conn закрывается и освобождается

            ' Конвертируем строки обратно в списки
            Dim originalList As List(Of Integer) = ConvertStringToList(originalStr)
            Dim sortedList As List(Of Integer) = ConvertStringToList(sortedStr)

            ' Возвращаем результат в виде кортежа
            Return New Tuple(Of List(Of Integer), List(Of Integer))(originalList, sortedList)

        Catch ex As Exception
            LogError($"Ошибка при загрузке записи с ID={id} из БД.", ex)
            Return Nothing ' Возвращаем Nothing при ошибке
        End Try
    End Function

    ''' <summary>
    ''' Вспомогательная функция для конвертации строки с числами через запятую в List(Of Integer).
    ''' </summary>
    ''' <param name="arrayString">Строка с числами, разделенными запятыми.</param>
    ''' <returns>Список List(Of Integer) или пустой список, если строка пуста или содержит ошибки.</returns>
    Private Function ConvertStringToList(arrayString As String) As List(Of Integer)
        Dim resultList As New List(Of Integer)()
        If String.IsNullOrWhiteSpace(arrayString) Then
            Return resultList ' Возвращаем пустой список, если строка пустая
        End If

        Dim parts() As String = arrayString.Split(","c) ' Разделяем по запятой
        For Each part As String In parts
            Dim number As Integer
            If Integer.TryParse(part.Trim(), number) Then ' Пытаемся преобразовать каждую часть в число
                resultList.Add(number)
            Else
                ' Логируем ошибку, если часть строки не является числом, но продолжаем обработку остальных
                LogError($"Ошибка конвертации '{part}' в число при загрузке массива из строки '{arrayString}'. Элемент пропущен.")
                ' Можно выбрать другую стратегию: вернуть Nothing или выбросить исключение,
                ' но пропуск элемента может быть более устойчивым вариантом.
            End If
        Next
        Return resultList
    End Function

    ' --- Добавьте также эту функцию для информационных сообщений ---
    ' Простой вспомогательный метод для логирования информационных сообщений
    Private Sub LogInfo(message As String, <CallerMemberName> Optional memberName As String = "")
        Dim infoMsg As String = $"[{DateTime.Now:G}] [{memberName}] {message}"

        ' Выводим в окно Output в Visual Studio
        Console.WriteLine("--- INFO ---")
        Console.WriteLine(infoMsg)
        Console.WriteLine("------------")
    End Sub

    ' Простой вспомогательный метод для логирования ошибок (можно расширить)
    Public Sub LogError(message As String, Optional ex As Exception = Nothing, <CallerMemberName> Optional memberName As String = "")
        Dim errorMsg As String = $"[{DateTime.Now:G}] [{memberName}] {message}"
        If ex IsNot Nothing Then
            errorMsg &= $"{vbCrLf}   Exception: {ex.GetType().Name} - {ex.Message}"

            ' --- Добавлено: Логирование вложенных исключений ---
            Dim innerEx As Exception = ex.InnerException
            Dim level As Integer = 1
            While innerEx IsNot Nothing
                errorMsg &= $"{vbCrLf}   InnerException (Level {level}): {innerEx.GetType().Name} - {innerEx.Message}"
                ' При необходимости можно добавить и StackTrace для InnerException
                ' errorMsg &= $"{vbCrLf}      Inner StackTrace: {innerEx.StackTrace}"
                innerEx = innerEx.InnerException
                level += 1
            End While
            ' --- Конец добавленного кода ---

            ' Основной StackTrace может быть очень длинным, оставим его опциональным или уберем пока
            ' errorMsg &= $"{vbCrLf}   StackTrace: {ex.StackTrace}"
        End If

        Console.WriteLine("--- ERROR ---")
        Console.WriteLine(errorMsg)
        Console.WriteLine("-------------")

        ' MessageBox.Show(message & If(ex IsNot Nothing, vbCrLf & ex.Message, ""), "Ошибка Базы Данных", MessageBoxButtons.OK, MessageBoxIcon.Error)
    End Sub

End Module

File: SourceGUI/DatabaseHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Data.Sqlite;
using System.IO;


namespace SourceGUI
{
    /// <summary>
    /// Представляет полные данные одной записи истории сортировки. Упрщеннные форматы для удобства ввода/вывода
    /// </summary>
    public class MainTableMetadata
    {
        public long ID { get; set; }
        public string Timestamp { get; set; } = string.Empty;
        public string Original { get; set; } = string.Empty;
        public string Sorted { get; set; } = string.Empty;
    }

    /// <summary>
    /// Представляет полные данные одной записи истории сортировки. Системные форматы для корректности данных
    /// </summary>
    public class MainTableEntry
    {
        public long ID { get; set; }
        public DateTime Timestamp { get; set; }
        public List<int> OriginalArray { get; set; } = new List<int>();
        public List<int> SortedArray { get; set; } = new List<int>();
    }


    /// <summary>
    /// Статический класс для работы с базой данных SQLite (MainDB.sqlite).
    /// </summary>
    public static class DatabaseHelper
    {
        private static string dbFileName = "MainDB.sqlite";

        /// <summary>
        /// Получает полный путь к файлу базы данных.
        /// БД располагается в директории запуска приложения.
        /// </summary>
        public static string GetDatabasePath()
        {
            return Path.Combine(AppContext.BaseDirectory, dbFileName);
        }

        /// <summary>
        /// Формирует строку подключения к SQLite.
        /// </summary>
        private static string GetConnectionString()
        {
            return $"Data Source={GetDatabasePath()}";
        }

        /// <summary>
        /// Гарантирует, что файл БД и таблица 'MainTable' существуют.
        /// Создает их при необходимости.
        /// </summary>
        /// <returns>True, если инициализация прошла успешно, иначе False.</returns>
        public static bool InitializeDatabase()
        {
            string sql = @"
                CREATE TABLE IF NOT EXISTS MainTable (
                    ID            INTEGER PRIMARY KEY,
                    Timestamp     TEXT NOT NULL,
                    OriginalArray TEXT NULL,
                    SortedArray   TEXT NULL
                );";

            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;
                        command.ExecuteNonQuery();
                    }
                }
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка инициализации БД: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Сохраняет результат сортировки в базу данных.
        /// </summary>
        /// <param name="originalArray">Исходный массив.</param>
        /// <param name="sortedArray">Отсортированный массив.</param>
        /// <returns>True при успехе, иначе False.</returns>
        public static bool SaveSortResult(List<int> originalArray, List<int> sortedArray)
        {
            if (originalArray == null || sortedArray == null) return false;

            string originalStr = string.Join(",", originalArray);
            string sortedStr = string.Join(",", sortedArray);
            string timestampStr = DateTime.UtcNow.ToString("o");
            string sql = @"INSERT INTO MainTable (ID, Timestamp, OriginalArray, SortedArray) VALUES (@ID, @Timestamp, @Original, @Sorted)";

            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();

                    // 1. Get the next ID
                    long nextId = 1; // Default ID for the first record
                    using (var command = new SqliteCommand("SELECT MAX(ID) FROM MainTable", connection))
                    {
                        object? result = command.ExecuteScalar();
                        if (result != null && result != DBNull.Value)
                        {
                            nextId = Convert.ToInt64(result) + 1;
                        }
                    }

                    // 2. Insert the new record
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;
                        command.Parameters.AddWithValue("@ID", nextId);
                        command.Parameters.AddWithValue("@Timestamp", timestampStr);
                        command.Parameters.AddWithValue("@Original", (object?)originalStr ?? DBNull.Value);
                        command.Parameters.AddWithValue("@Sorted", (object?)sortedStr ?? DBNull.Value);

                        command.ExecuteNonQuery();
                    }
                }
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка сохранения в БД: {ex.Message}");
                return false;
            }
        }


        /// <summary>
        /// Получает метаданные всех записей истории для отображения в гриде.
        /// </summary>
        /// <returns>Список метаданных или пустой список при ошибке/отсутствии данных.</returns>
        public static List<MainTableMetadata> GetAllMainTableMetadata()
        {
            var historyList = new List<MainTableMetadata>();
            string sql = "SELECT * FROM MainTable ORDER BY Timestamp DESC"; // Сортируем по убыванию даты

            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;
                        using (var reader = command.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                string timestampStr = reader.GetString(1);
                                DateTime timestamp = DateTime.Parse(timestampStr).ToLocalTime();
                                string formattedTimestamp = timestamp.ToString("yyyy-MM-dd HH:mm:ss");

                                historyList.Add(new MainTableMetadata
                                {
                                    ID = reader.GetInt64(0), // ID - INTEGER PRIMARY KEY -> long
                                    Timestamp = formattedTimestamp,
                                    Original = reader.GetString(2),
                                    Sorted = reader.GetString(3),
                                });
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка чтения истории из БД: {ex.Message}");
                // Возвращаем пустой список в случае ошибки
            }
            return historyList;
        }

        /// <summary>
        /// Загружает полную запись истории (с массивами) по её ID.
        /// </summary>
        /// <param name="id">ID записи.</param>
        /// <returns>Объект MainTableEntry или null, если запись не найдена или произошла ошибка.</returns>
        public static MainTableEntry? GetSortResultById(long id)
        {
            MainTableEntry? entry = null;
            string sql = "SELECT Timestamp, OriginalArray, SortedArray FROM MainTable WHERE ID = @ID";

            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;
                        command.Parameters.AddWithValue("@ID", id);

                        using (var reader = command.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                string timestampStr = reader.GetString(0);
                                string originalStr = reader.IsDBNull(1) ? string.Empty : reader.GetString(1);
                                string sortedStr = reader.IsDBNull(2) ? string.Empty : reader.GetString(2);

                                entry = new MainTableEntry
                                {
                                    ID = id,
                                    Timestamp = DateTime.Parse(timestampStr).ToLocalTime(), // Преобразуем в DateTime
                                    OriginalArray = ParseIntListFromString(originalStr),
                                    SortedArray = ParseIntListFromString(sortedStr)
                                };
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка загрузки записи {id} из БД: {ex.Message}");
            }
            return entry;
        }

        /// <summary>
        /// Обновляет существующую запись в базе данных.
        /// </summary>
        /// <param name="id">ID записи для обновления.</param>
        /// <param name="originalArray">Новый исходный массив.</param>
        /// <param name="sortedArray">Новый отсортированный массив.</param>
        /// <returns>True при успехе, иначе False.</returns>
        public static bool UpdateSortResult(long id, List<int> originalArray, List<int> sortedArray)
        {
            if (originalArray == null || sortedArray == null) return false;

            string originalStr = string.Join(",", originalArray);
            string sortedStr = string.Join(",", sortedArray);
            // Можно обновить и Timestamp, если это требуется по логике
            string timestampStr = DateTime.UtcNow.ToString("o");

            string sql = @"
                UPDATE MainTable
                SET Timestamp = @Timestamp,
                    OriginalArray = @Original,
                    SortedArray = @Sorted
                WHERE ID = @ID";

            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;
                        command.Parameters.AddWithValue("@Timestamp", timestampStr);
                        command.Parameters.AddWithValue("@Original", (object?)originalStr ?? DBNull.Value);
                        command.Parameters.AddWithValue("@Sorted", (object?)sortedStr ?? DBNull.Value);
                        command.Parameters.AddWithValue("@ID", id);

                        int rowsAffected = command.ExecuteNonQuery();
                        return rowsAffected > 0; // Успех, если хотя бы одна строка была обновлена
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка обновления записи {id} в БД: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Вспомогательный метод для парсинга строки с числами через запятую в список int.
        /// </summary>
        private static List<int> ParseIntListFromString(string? data)
        {
            var list = new List<int>();
            if (string.IsNullOrWhiteSpace(data))
            {
                return list; // Возвращаем пустой список, если строка пуста или null
            }

            string[] parts = data.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string part in parts)
            {
                if (int.TryParse(part.Trim(), out int number))
                {
                    list.Add(number);
                }
                else
                {
                    Console.WriteLine($"Предупреждение: не удалось распарсить элемент '{part}' в число.");
                }
            }
            return list;
        }

        /// <summary>
        /// Получает одну случайную полную запись из истории.
        /// </summary>
        /// <returns>Объект MainTableEntry или null, если таблица пуста или произошла ошибка.</returns>
        public static MainTableEntry? GetRandomSortResult()
        {
            MainTableEntry? entry = null;
            // Используем ORDER BY RANDOM() LIMIT 1 для получения случайной строки в SQLite
            string sql = "SELECT ID, Timestamp, OriginalArray, SortedArray FROM MainTable ORDER BY RANDOM() LIMIT 1";

            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;

                        using (var reader = command.ExecuteReader())
                        {
                            if (reader.Read()) // Если запись найдена
                            {
                                long id = reader.GetInt64(0);
                                string timestampStr = reader.GetString(1);
                                string originalStr = reader.IsDBNull(2) ? string.Empty : reader.GetString(2);
                                string sortedStr = reader.IsDBNull(3) ? string.Empty : reader.GetString(3);

                                entry = new MainTableEntry
                                {
                                    ID = id,
                                    Timestamp = DateTime.Parse(timestampStr).ToLocalTime(),
                                    OriginalArray = ParseIntListFromString(originalStr),
                                    SortedArray = ParseIntListFromString(sortedStr)
                                };
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка получения случайной записи из БД: {ex.Message}");
                // Возвращаем null в случае ошибки
            }
            return entry;
        }

        /// <summary>
        /// Получает количество записей в таблице MainTable.
        /// </summary>
        /// <returns>Количество записей или -1 при ошибке.</returns>
        public static long GetHistoryCount()
        {
            string sql = "SELECT COUNT(*) FROM MainTable";
            try
            {
                 using (var connection = new SqliteConnection(GetConnectionString()))
                 {
                    connection.Open();
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = sql;
                        // ExecuteScalar возвращает первое поле первой строки (или null)
                        object? result = command.ExecuteScalar();
                        if (result != null && result != DBNull.Value)
                        {
                            return Convert.ToInt64(result);
                        }
                    }
                 }
                 return 0; // Если таблица пуста, COUNT вернет 0
            }
            catch (Exception ex)
            {
                 Console.WriteLine($"Ошибка получения количества записей из БД: {ex.Message}");
                 return -1; // Признак ошибки
            }
        }

        public static bool DeleteSortResult(long id)
        {
            string deleteSql = "DELETE FROM MainTable WHERE ID = @Id;";
            string updateSql = "UPDATE MainTable SET ID = ID - 1 WHERE ID > @Id;";
            try
            {
                using (var connection = new SqliteConnection(GetConnectionString()))
                {
                    connection.Open();
                    using (var transaction = connection.BeginTransaction()) // Используем транзакцию для целостности
                    {
                        using (var deleteCommand = new SqliteCommand(deleteSql, connection, transaction))
                        {
                            deleteCommand.Parameters.AddWithValue("@Id", id);
                            deleteCommand.ExecuteNonQuery();
                        }
                        using (var updateCommand = new SqliteCommand(updateSql, connection, transaction))
                        {
                            updateCommand.Parameters.AddWithValue("@Id", id);
                            updateCommand.ExecuteNonQuery();
                        }

                        transaction.Commit(); // Подтверждаем транзакцию, если обе операции прошли успешно
                    }
                    connection.Close(); // Явно закрываем соединение
                }
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка при удалении записи с ID={id}: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Проверяет доступность соединения с БД.
        /// </summary>
        /// <returns>True, если соединение успешно открыто и закрыто.</returns>
        public static bool CheckDbConnection()
        {
            try
            {
                 using (var connection = new SqliteConnection(GetConnectionString()))
                 {
                     connection.Open();
                     return true; // Соединение успешно
                 }
            }
            catch (Exception)
            {
                 return false; // Ошибка соединения
            }
        }
    }
}
